pkg/sanity/cleanup.go

```diff
@@ -36,110 +42,284 @@ type VolumeInfo struct {
+	}
+	return vol, err
 }
 
-// RegisterVolume adds or updates an entry for the volume with the
-// given name.
-func (cl *Cleanup) RegisterVolume(name string, info VolumeInfo) {
+func (cl *Cleanup) deleteVolume(ctx context.Context, req *csi.DeleteVolumeRequest) (*csi.DeleteVolumeResponse, error) {
+	vol, err := cl.ControllerClient.DeleteVolume(ctx, req)
+	if err == nil {
+		cl.unregisterVolume(req.VolumeId)
+	}
+	return vol, err
+}
+
+// MustControllerPublishVolume is like ControllerPublishVolume but asserts that
+// the volume was successfully controller-published.
+func (cl *Cleanup) MustControllerPublishVolume(ctx context.Context, req *csi.ControllerPublishVolumeRequest) *csi.ControllerPublishVolumeResponse {
+	conpubvol, err := cl.controllerPublishVolume(ctx, req)
+	Expect(err).NotTo(HaveOccurred())
+	Expect(conpubvol).NotTo(BeNil())
+	return conpubvol
+}
+
+func (cl *Cleanup) controllerPublishVolume(ctx context.Context, req *csi.ControllerPublishVolumeRequest) (*csi.ControllerPublishVolumeResponse, error) {
+	conpubvol, err := cl.ControllerClient.ControllerPublishVolume(ctx, req)
+	if err == nil && req.VolumeId != "" && req.NodeId != "" {
+		cl.registerVolume(VolumeInfo{VolumeID: req.VolumeId, NodeID: req.NodeId})
+	}
+	return conpubvol, err
+}
+
+// registerVolume adds or updates an entry for given volume.
+func (cl *Cleanup) registerVolume(info VolumeInfo) {
+	Expect(info).NotTo(BeNil())
+	Expect(info.VolumeID).NotTo(BeEmpty())
 	cl.mutex.Lock()
 	defer cl.mutex.Unlock()
 	if cl.volumes == nil {
-		cl.volumes = make(map[string]VolumeInfo)
+		cl.volumes = make(map[string]*VolumeInfo)
 	}
-	cl.volumes[name] = info
+	cl.volumes[info.VolumeID] = &info
 }
 
-// MaybeRegisterVolume adds or updates an entry for the volume with
-// the given name if CreateVolume was successful.
-func (cl *Cleanup) MaybeRegisterVolume(name string, vol *csi.CreateVolumeResponse, err error) {
-	if err == nil && vol.GetVolume().GetVolumeId() != "" {
-		cl.RegisterVolume(name, VolumeInfo{VolumeID: vol.GetVolume().GetVolumeId()})
+// unregisterVolume removes the entry for the volume with the
+// given ID, thus preventing all cleanup operations for it.
+func (cl *Cleanup) unregisterVolume(id string) {
+	cl.mutex.Lock()
+	defer cl.mutex.Unlock()
+	cl.unregisterVolumeNoLock(id)
+}
+
+func (cl *Cleanup) unregisterVolumeNoLock(id string) {
+	Expect(id).NotTo(BeEmpty())
+	if cl.volumes != nil {
+		delete(cl.volumes, id)
 	}
 }
 
-// UnregisterVolume removes the entry for the volume with the
-// given name, thus preventing all cleanup operations for it.
-func (cl *Cleanup) UnregisterVolume(name string) {
+// MustCreateSnapshot is like CreateSnapshot but asserts that the snapshot was
+// successfully created.
+func (cl *Cleanup) MustCreateSnapshot(ctx context.Context, req *csi.CreateSnapshotRequest) *csi.CreateSnapshotResponse {
+	snap, err := cl.createSnapshot(ctx, req)
+	Expect(err).NotTo(HaveOccurred())
```

Comment:
A problem with assertions in helper functions is that Ginkgo only reports one source line for the error by default; even with `-trace`, that initial line is typically not very informative.

Better use `ExpectWithOffset` and add an offset parameter to the `MustCreateSnapshot` parameters so that this helper function an also be called indirectly through some other helper functions.

Also, `NotTo(HaveOccurred())` without additional explanation is potentially problematic, depending on how much information is in the error. Much too often the error is very generic, in which case the assertion produced by Gomega doesn't say anything about what failed.

Better always use `NoTo(HaveOccurred(), "create snapshot"`, potentially even with further parameters.

I know, much of the existing code doesn't do that properly either :cry:

Reply:
Ah yes, I remember you previously mentioned how `To()` should have a description but forgot about it again. I made sure it's now set.

I also updated the code to specify and/or pass through the offset everywhere. It's not a beauty though, I wonder if ginkgo could do better here by deriving the offset automatically (at least after the top-level `t.Helper()`-like indicator).